src/p2p/swarm.rs

```diff
@@ -335,53 +347,35 @@ impl NetworkBehaviour for SwarmApi {
-            // for soon.
-            self.events
-                .push_back(swarm::NetworkBehaviourAction::DialPeer {
-                    peer_id: *peer_id,
-                    condition: DialPeerCondition::NotDialing,
-                });
-        }
-
-        // this should not be executed once, but probably will be in case unsupported addresses or something
-        // surprising happens.
-        for failed in self.pending_connections.remove(peer_id).unwrap_or_default() {
-            self.connect_registry
-                .finish_subscription(failed.into(), Err("addresses exhausted".into()));
-        }
-    }
-
-    fn inject_addr_reach_failure(
+    fn inject_dial_failure(
         &mut self,
-        peer_id: Option<&PeerId>,
-        addr: &Multiaddr,
-        error: &dyn std::error::Error,
+        peer_id: Option<PeerId>,
+        _handler: Self::ConnectionHandler,
+        error: &DialError,
     ) {
-        trace!("inject_addr_reach_failure {} {}", addr, error);
-
+        trace!("inject_dial_failure: {:?} ({})", peer_id, error);
         if let Some(peer_id) = peer_id {
-            match self.pending_connections.entry(*peer_id) {
-                Entry::Occupied(mut oe) => {
-                    let addresses = oe.get_mut();
-                    let addr = MultiaddrWithPeerId::try_from(addr.clone())
-                        .expect("dialed address contains peerid in libp2p 0.38");
-                    let pos = addresses.iter().position(|a| *a == addr);
-
-                    if let Some(pos) = pos {
-                        addresses.swap_remove(pos);
-                        self.connect_registry
-                            .finish_subscription(addr.into(), Err(error.to_string()));
-                    }
+            if self.pending_addresses.contains_key(&peer_id) {
+                // it is possible that these addresses have not been tried yet; they will be asked
+                // for soon.
+                let handler = self.new_handler();
+                self.events.push_back(swarm::NetworkBehaviourAction::Dial {
+                    opts: DialOpts::peer_id(peer_id)
+                        .condition(PeerCondition::NotDialing)
+                        .build(),
+                    handler,
+                });
+            }
 
-                    if addresses.is_empty() {
-                        oe.remove();
-                    }
-                }
-                Entry::Vacant(_) => {}
+            // this should not be executed once, but probably will be in case unsupported addresses or something
+            // surprising happens.
+            for failed in self
+                .pending_connections
+                .remove(&peer_id)
+                .unwrap_or_default()
+            {
+                self.connect_registry
+                    .finish_subscription(failed.into(), Err("addresses exhausted".into()));
             }
```

Comment:
Doing this unconditionally creates the failure for `p2p::swarm::tests::racy_connecting_attempts`.

I wonder ... why were changes around this necessary? Seems this logic would had been easiest to do handle with the old structure using the entry api?